tooi/api/streaming.py

```diff
@@ -0,0 +1,392 @@
+
+    async def nsubscribers(self):
+        assert self.lock.locked()
+        return len(self.subscribers)
+
+    async def _run(self):
+        while True:
+            e = await self.client.queue.get()
+            async with self.lock:
+                for subscriber in self.subscribers:
+                    await subscriber.dispatch(e)
+
+    async def close(self):
+        assert self.lock.locked()
+        self.client_task.cancel()
+        self.client_task = None
+
+        self.mplx_task.cancel()
+        self.mplx_task = None
+
+        await self.client.close()
+        self.client = None
+
+    async def add_subscriber(self) -> StreamSubscription:
+        assert self.lock.locked()
+        subscription = StreamSubscription(self.mplx, self.stream)
+
+        logger.info(
+                (f"StreamInstance: stream={self.client.stream_name} "
+                 f"has {await self.nsubscribers()} refs now"))
+        self.subscribers.add(subscription)
+
+        return subscription
+
+    async def remove_subscriber(self, subscriber):
+        assert self.lock.locked()
+        self.subscribers.remove(subscriber)
+
+
+class StreamMultiplexer:
+    """
+    Internal class for multiplexing multiple streams into a single queue.  Call open_stream() to
+    start receiving events on a stream, and close_stream() to stop receiving them.  The number of
+    listeners for each stream type is tracked, and the stream won't be closed unless close_stream()
+    is called the same number of times open_stream() was called.
+    """
+
+    # lock order: self.lock -> StreamInstance.lock
+    #         or: StreamInstance.lock if self.lock not held
+    #        NOT: StreamInstance.lock -> self.lock
+
+    def __init__(self, instance: InstanceInfo):
+        self.instance = instance
+        self.streams = {}
+        self.lock = asyncio.Lock()
+
+    async def open_stream(self, stream: str) -> StreamSubscription:
+        async with self.lock:
+            if (sinst := self.streams.get(stream)) is None:
+                logger.info(f"StreamMultiplexer: new client for stream={stream}")
+                sinst = StreamInstance(self, self.instance, stream)
+                self.streams[stream] = sinst
+
+            async with sinst.lock:
+                return await sinst.add_subscriber()
+
+    async def close_stream(self, subscription: StreamSubscription):
+        async with self.lock:
+            if (sinst := self.streams.get(subscription.stream)) is None:
+                raise (KeyError("the subscription's stream was not found"))
+
+            stream = subscription.stream
+
+            async with sinst.lock:
+                await sinst.remove_subscriber(subscription)
+
+                if await sinst.nsubscribers() > 0:
+                    return
+
+                await sinst.close()
+
+            del self.streams[stream]
+
+    def run(self):
+        self._task = run_async_task(self._run())
```

Comment:
I don't think this is used, since `StreamMultiplexer` doesn't have a `_run` method.

Reply:
fixed in a9e6d8f